### PR TITLE
feat(licenseinfo): Generate different variants of DOCX files

### PR DESCRIPTION
**Pull request by maierthomas**
this PR supports to generate different output variants for the DOCX generator
- disclosure license info file
- report license info file (contains sw360 license todos)
- add project title (project name and version) to text/xhtml generator

licenseinfo-generators

### CODE REVIEW

**Comment by adrian-evo** [comment]
Works as expected.

**Comment by maxhbr** [comment]
In the long term: maybe implement serviceloading, to get the code moudlarized.

File: backend/src/src-licenseinfo/src/main/java/org/eclipse/sw360/licenseinfo/outputGenerators/DocxGenerator.java [hunk 1]
```diff
@@ -121,6 +132,44 @@ private void addReleaseTitle(XWPFDocument document, LicenseInfoParsingResult par
...
+            addTodos(document, licenseName, sw360Licenes);
+        }
+    }
+
+    private void addTodos(XWPFDocument document, String spdxLicense, List<License> sw360Licenses) {
+        if (sw360Licenses != null && !sw360Licenses.isEmpty()) {
+            XWPFRun todoTitleRun = document.createParagraph().createRun();
+            addNewLines(todoTitleRun, 0);
+            Set<String> todos = getTodosFromLicenses(spdxLicense, sw360Licenses);
+            addFormattedText(todoTitleRun, "Todos for license " + spdxLicense + ":", FONT_SIZE, true);
```
[the remark this answers is not in the document]
Comment (maierthomas) [reply backend/src/src-licenseinfo/src/main/java/org/eclipse/sw360/licenseinfo/outputGenerators/DocxGenerator.java]: renamed text and function name to obligations

File: backend/src/src-licenseinfo/src/main/java/org/eclipse/sw360/licenseinfo/outputGenerators/DocxGenerator.java [hunk 2]
```diff
@@ -75,32 +96,22 @@ private void fillReleaseBulletList(XWPFDocument document, Collection<LicenseInfo
...
-        setText(releaseInfoTextRun, "Please note the following license conditions and copyright " +
+    private void fillReleaseDetailList(XWPFDocument document, Collection<LicenseInfoParsingResult> projectLicenseInfoResults, boolean todos) {
+        addFormattedText(document.createParagraph().createRun(), "Detailed Releases Information", FONT_SIZE + 2, true);
+        setText(document.createParagraph().createRun(), "Please note the following license conditions and copyright " +
                 "notices applicable to Open Source Software and/or other components (or parts thereof):");
         addNewLines(document, 0);
 
+        List<License> sw360Licenses = new ArrayList<>();
+        if (todos) {
+            sw360Licenses = getLicenses();
```
[the remark this answers is not in the document]
Comment (maierthomas) [reply backend/src/src-licenseinfo/src/main/java/org/eclipse/sw360/licenseinfo/outputGenerators/DocxGenerator.java]: moved logic and renamed the parameter

File: backend/src/src-licenseinfo/src/main/java/org/eclipse/sw360/licenseinfo/outputGenerators/DocxGenerator.java [hunk 3]
```diff
@@ -154,6 +203,33 @@ private String getFilename(LicenseInfoParsingResult licenseInfoParsingResult) {
...
+        return todos;
+    }
+
+    private static List<License> getLicenses() {
+        LicenseService.Iface licenseClient = new ThriftClients().makeLicenseClient();
+        try {
+            return licenseClient.getLicenses();
+        } catch (TException e) {
+            LOGGER.error("Could not get license: " + e.getMessage(), e);
+            return null;
```
[the remark this answers is not in the document]
Comment (maierthomas) [reply backend/src/src-licenseinfo/src/main/java/org/eclipse/sw360/licenseinfo/outputGenerators/DocxGenerator.java]: To provide a document is even better as to stop the generation.
Added alert text if the output generator could not load the sw360 licenses.

File: backend/src/src-licenseinfo/src/main/java/org/eclipse/sw360/licenseinfo/outputGenerators/OutputGenerator.java [hunk 4]
```diff
@@ -102,9 +108,9 @@ public VelocityContext getConfiguredVelocityContext() {
     }
 
     @NotNull
-    protected LicenseInfoParsingResult mergeLicenseInfoParsingResults(LicenseInfoParsingResult r1, LicenseInfoParsingResult r2){
+    protected LicenseInfoParsingResult mergeLicenseInfoParsingResults(LicenseInfoParsingResult r1, LicenseInfoParsingResult r2) {
```
[the remark this answers is not in the document]
Comment (maierthomas) [reply backend/src/src-licenseinfo/src/main/java/org/eclipse/sw360/licenseinfo/outputGenerators/OutputGenerator.java]: removed unnecessary formattings

File: backend/src/src-licenseinfo/src/main/java/org/eclipse/sw360/licenseinfo/outputGenerators/DocxGenerator.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (maxhbr) [reply backend/src/src-licenseinfo/src/main/java/org/eclipse/sw360/licenseinfo/outputGenerators/DocxGenerator.java]: This does not help if the document ist generated / shipped automatically or if it is never opened (what is rather likely).

File: backend/src/src-licenseinfo/src/main/java/org/eclipse/sw360/licenseinfo/outputGenerators/DocxGenerator.java [hunk 5]
```diff
@@ -154,6 +203,33 @@ private String getFilename(LicenseInfoParsingResult licenseInfoParsingResult) {
                 .orElse(UNKNOWN_FILE_NAME);
     }
 
+    private static Set<String> getTodosFromLicenses(String spdxLicense, List<License> licenses) {
+        Set<String> todos = new HashSet<>();
+        if (!spdxLicense.isEmpty() && !Objects.isNull(licenses)) {
+            for (License license : licenses) {
+                if (spdxLicense.equalsIgnoreCase(license.getId())) {
+                    for (Todo todo : license.getTodos()) {
+                        todos.add(todo.getText());
```
[the remark this answers is not in the document]
Comment (maierthomas) [reply backend/src/src-licenseinfo/src/main/java/org/eclipse/sw360/licenseinfo/outputGenerators/DocxGenerator.java]: The obligations are in relation to the user, aren't they?
Which information of obligations should we print to the document?

[the remark this answers is not in the document]
Comment (maxhbr) [reply backend/src/src-licenseinfo/src/main/java/org/eclipse/sw360/licenseinfo/outputGenerators/DocxGenerator.java]: No, the todos depend on the viewing user. The Obligations are only chained to that.

File: backend/src/src-licenseinfo/src/main/java/org/eclipse/sw360/licenseinfo/outputGenerators/DocxGenerator.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (maierthomas) [reply backend/src/src-licenseinfo/src/main/java/org/eclipse/sw360/licenseinfo/outputGenerators/DocxGenerator.java]: Ok - the generator fails now if the sw360 license service could not determine the licenses.

**Comment by maxhbr** [comment]
@maierthomas related to the comment https://github.com/sw360/sw360portal/pull/726#discussion_r171606264: Is the report currently user / organization dependent in its list of todos?

**Comment by maierthomas** [comment]
You mean we should get the todo information over the `LicenseDatabaseHandler` instead of calling the thrift service for licenses directly?

**Comment by maxhbr** [comment]
@maierthomas related to the comment https://github.com/sw360/sw360portal/pull/726#discussion_r171610932: but there are still other places where the errors are just logged to the document? So it would be more consistent to do this also here and just clearify that the document needs manual review. If one wants to have automatic generation for automatic shipping, it would the be an option to implement an configurable error handler.

**Comment by maierthomas** [comment]
yes but this is related to pull request #719 .
We want to generate the report document in case of a licenseinfo parsing failure.

**Comment by maierthomas** [comment]
Output generators should response with a result value (0 = success, 1 = parsing failure etc...) -> new issue. This pull request should not contain this feature.

@mcjaeger should the reports contain all todos of a license (not filtered by the organization)?

**Comment by mcjaeger** [comment]
from what I understood all todos for the first implementation would be sufficient.

**Comment by mcjaeger** [comment]
@maxhbr I agree that your proposed error handling would be better.

But this is not intended part of this PR as this is the same behavior for the existing three reports.

**Comment by mcjaeger** [comment]
@maxhbr could you pls. clarify if other requested changes are covered?

**Pull request merged**